### PR TITLE
Add labels to mipsy_web data view

### PR DESCRIPTION
**Pull request by Dylan-Brotherston**

### CODE REVIEW

**Comment by joooooooooooooooooooooooooooooooooooosh** [comment]
doesn't need to be in this PR but adding a button somewhere to toggle the labels showing in the rendered data segment itself rather than just hovering on tooltip would be nice

**Review by joooooooooooooooooooooooooooooooooooosh — APPROVED** [review]

File: crates/mipsy_lib/src/inst/register.rs [hunk 1]
```diff
@@ -40,16 +40,37 @@ pub enum Register {
...
+    Register::S5,
+    Register::S6,
+    Register::S7,
+    Register::T8,
+    Register::T9,
+    Register::K0,
+    Register::K1,
+    Register::Gp,
+    Register::Sp,
+    Register::Fp,
```
Comment (joooooooooooooooooooooooooooooooooooosh) [inline crates/mipsy_lib/src/inst/register.rs:73]: I would rather this be left in the more compact (and organised) original form, you can add `#[rustfmt::skip]` to line 41 to preserve the formatting when running `rustfmt` on everything else

File: crates/mipsy_lib/src/inst/register.rs [hunk 2]
```diff
@@ -107,55 +125,55 @@ impl Register {
...
+            Self::S6 => 22,
+            Self::S7 => 23,
+            Self::T8 => 24,
+            Self::T9 => 25,
+            Self::K0 => 26,
+            Self::K1 => 27,
+            Self::Gp => 28,
+            Self::Sp => 29,
+            Self::Fp => 30,
+            Self::Ra => 31,
```
Comment (joooooooooooooooooooooooooooooooooooosh) [inline crates/mipsy_lib/src/inst/register.rs:161]: same here

File: crates/mipsy_lib/src/inst/register.rs [hunk 3]
```diff
@@ -107,55 +125,55 @@ impl Register {
...
+            0 => Ok(Self::Zero),
+            1 => Ok(Self::At),
+            2 => Ok(Self::V0),
+            3 => Ok(Self::V1),
+            4 => Ok(Self::A0),
+            5 => Ok(Self::A1),
+            6 => Ok(Self::A2),
+            7 => Ok(Self::A3),
+            8 => Ok(Self::T0),
+            9 => Ok(Self::T1),
```
Comment (joooooooooooooooooooooooooooooooooooosh) [inline crates/mipsy_lib/src/inst/register.rs:176]: and here (the new formatting for `_` case is fine)

File: crates/mipsy_lib/src/inst/register.rs [hunk 4]
```diff
@@ -200,77 +214,77 @@ impl Register {
...
+            Self::S6 => "S6",
+            Self::S7 => "S7",
+            Self::T8 => "T8",
+            Self::T9 => "T9",
+            Self::K0 => "K0",
+            Self::K1 => "K1",
+            Self::Gp => "GP",
+            Self::Sp => "SP",
+            Self::Fp => "FP",
+            Self::Ra => "RA",
```
Comment (joooooooooooooooooooooooooooooooooooosh) [inline crates/mipsy_lib/src/inst/register.rs:250]: here

File: crates/mipsy_lib/src/inst/register.rs [hunk 5]
```diff
@@ -200,77 +214,77 @@ impl Register {
...
+            Self::S6 => "s6",
+            Self::S7 => "s7",
+            Self::T8 => "t8",
+            Self::T9 => "t9",
+            Self::K0 => "k0",
+            Self::K1 => "k1",
+            Self::Gp => "gp",
+            Self::Sp => "sp",
+            Self::Fp => "fp",
+            Self::Ra => "ra",
```
Comment (joooooooooooooooooooooooooooooooooooosh) [inline crates/mipsy_lib/src/inst/register.rs:287]: here

File: crates/mipsy_web/src/components/data_segment.rs [hunk 6]
```diff
@@ -145,8 +159,8 @@ fn render_page(props: &DataSegmentProps, page_addr: u32, page_contents: Vec<Safe
                         let full_offset = nth * ROW_SIZE + offset;
                         let full_page_addr = page_addr as usize + full_offset;
                         let class = "cursor-help";
-                        let mut style = "border: 2px solid ".to_string();
-                        let mut title = "".to_string();
+                        let mut style = String::new();
```
Comment (joooooooooooooooooooooooooooooooooooosh) [inline crates/mipsy_web/src/components/data_segment.rs:162]: style could still be initialised as `String::from("border: 2px solid ");` but that's up to you

**Review by joooooooooooooooooooooooooooooooooooosh — CHANGES_REQUESTED** [review]

**Review by joooooooooooooooooooooooooooooooooooosh — APPROVED** [review]

**Comment by Dylan-Brotherston** [comment]
This has a massive performance overhead and might want to be rolled back and rethought.

**Comment by joooooooooooooooooooooooooooooooooooosh** [comment]
wait really? can you give examples of where this is noticeable? The only thing added I can see that could have a noticeable perf impact is the label iteration on every byte, but this could be rethought easily by creating a HashMap once before the loops start

**Comment by Dylan-Brotherston** [comment]
It's hard to do an exact benchmark. 
But it is noticeably slower.

If you are actively viewing the data tab it is extremely obvious it has slowed down.

**Comment by joooooooooooooooooooooooooooooooooooosh** [comment]
I'm not noticing anything? Stepping through commands that modify the data tab and/or position of $sp/$fp affect the data tab instantaneously and even running two_powerful to completion updates without issue.

**Pull request merged**